cpp/src/arrow/array/concatenate.cc

```diff
@@ -243,35 +243,57 @@ class ConcatenateImpl {
  private:
   // Gather the index-th buffer of each input into a vector.
   // Bytes are sliced with that input's offset and length.
+  // Note that BufferVector will not contain the buffer of in_[i] if it's
+  // nullptr.
   BufferVector Buffers(size_t index) {
-    BufferVector buffers(in_.size());
+    BufferVector buffers;
+    buffers.reserve(in_.size());
     for (size_t i = 0; i < in_.size(); ++i) {
-      buffers[i] = SliceBuffer(in_[i].buffers[index], in_[i].offset, in_[i].length);
+      const auto& buffer = in_[i].buffers[index];
+      if (buffer) {
+        buffers.push_back(
+            SliceBuffer(in_[i].buffers[index], in_[i].offset, in_[i].length));
```

Comment:
(optional, not your change) not really your change but it seems like we could use the iterator pattern above to avoid using in_[i] in multiple places.